[HUDI-1951] Add bucket hash index, compatible with the hive bucket

minihippo opened the pull request:
## What is the purpose of the pull request

Index pattern 1 in `RFC-29: Hash Index`
https://cwiki.apache.org/confluence/display/HUDI/RFC+-+29%3A+Hash+Index

## Brief change log

  - add a new index method
  - fix the compatibility issue when adding  member variables to HoodieKey

## Verify this pull request

This change added tests and can be verified as follows:
  - *Added unit tests for bucket index and serialization*
  - *Modify existing integration tests to verify the new index method*


## Committer checklist

 - [ ] Has a corresponding JIRA in PR title & commit
 
 - [ ] Commit message is descriptive of the change
 
 - [ ] CI is green

 - [ ] Necessary doc changes done or have another open PR
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

minihippo commented:
The `ITTestHoodieDemo` failure is due to its own instability rather than being affected by this new feature, and HUDI-2113 has already fixed it.

vinothchandar commented:
@minihippo can you please rebase this PR again?

minihippo commented:
@vinothchandar done

vinothchandar commented:
@minihippo been behind on this. 

is there anyone else who wants to take a first pass?

leesf commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieIndexConfig.java:
```diff
@@ -189,6 +189,13 @@
       .defaultValue("false")
       .withDocumentation("");
 
+  // ***** Hive Bucket Index Configs *****
```
can we replace `Hive` with `Hudi`?

minihippo replied:
done

leesf commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/keygen/SimpleAvroKeyGenerator.java:
```diff
@@ -30,19 +33,36 @@
         ? Collections.emptyList()
         : Collections.singletonList(recordKeyField);
     this.partitionPathFields = Collections.singletonList(partitionPathField);
+    if (!StringUtils.isNullOrEmpty(indexKeyField) && !indexKeyField.equals(recordKeyField)) {
```
`indexKeyField` must be equals to `recordKeyField` but not subset?

minihippo replied:
Incorrect check here. But for the bucket index, indexKeyField can be the subset. There is an one2one match between bucketId and file groupId. Therefore, the record indexed by `colA` is always clustered to the same bucket and updated by key `colA` and `colB` with the old one stored in the bucket

leesf commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/commit/BucketInfo.java:
```diff
@@ -30,6 +30,10 @@
   String partitionPath;
 
+  public BucketInfo() {
+    new BucketInfo(null, null, null);
```
remove this line?

minihippo replied:
done

leesf commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/utils/HiveBucketUtils.java:
```diff
@@ -0,0 +1,91 @@
+
+import org.apache.hudi.common.fs.FSUtils;
+
+public class HiveBucketUtils {
```
HoodieBucketUtils?

minihippo replied:
done

leesf commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/utils/HiveBucketUtils.java:
```diff
@@ -0,0 +1,91 @@
+  }
+
+  // copied from spark HiveHashFunction class
+  // see https://code.byted.org/bytedance/spark/blob/branch-3.0-bd/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala
```
please remove the internal link.

minihippo replied:
done

leesf commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/utils/HiveBucketUtils.java:
```diff
@@ -0,0 +1,91 @@
+    return mod(hash & Integer.MAX_VALUE, numBuckets);
+  }
+
+  // uuid 00499982-0f0d-473e-965d-e8a7476ec429
```
remove this line?

minihippo replied:
done

leesf commented on hudi-client/hudi-client-common/src/test/resources/hive_bucket_id_check.csv:
```diff
@@ -0,0 +1,5017 @@
+true,4329,733532,0.6628716966238256,0.7140616198749685,23
+false,4330,1956003,0.7879164384812248,0.7871551087042735,38
+false,4331,1239587,0.6565996423539073,0.10640163386619472,24
+true,4332,380900,0.860977145579482,0.150321088560726
```
any reason we need so much items in this file?

minihippo replied:
select the top 10 records as test case

minihippo commented:
Hi @leesf, I consider the patch is too large. Should I divided it into 2 pr for better review?

leesf commented:
> Hi @leesf, I consider the patch is too large. Should I divided it into 2 pr for better review?

After removing the csv file, the PR become smaller and i think it is ok to contains the changes in one PR.

leesf commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/keygen/BaseKeyGenerator.java:
```diff
@@ -71,11 +84,25 @@ public final HoodieKey getKey(GenericRecord record) {
   }
+
+  protected void validateIndexKeyField() {
+    if (indexKeyFields.size() != 0) {
```
would indexKeyFields be empty or mandatory when using hash index?

minihippo replied:
Mandatory. If empty, will throw a hoodieException with related info.

leesf replied:
so here if indexKeyFields is null or empty, should throw exception here?

leesf replied:
and please add uts to cover the changes.

minihippo replied:
Check the bucket index key field config when building the HoodieIndexConfig and add the relevant uts in `TestHoodieIndex`

leesf commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/commit/BaseCommitActionExecutor.java:
```diff
@@ -75,6 +77,11 @@ public BaseCommitActionExecutor(HoodieEngineContext context, HoodieWriteConfig c
+    // TODO: HUDI-2155 bulk insert support bucket index, HUDI-2156 cluster the table with bucket index.
+    if (this.config.getIndexType() == HoodieIndex.IndexType.BUCKET_INDEX
+        && (operationType == WriteOperationType.BULK_INSERT || operationType == WriteOperationType.CLUSTER)) {
+      throw new HoodieIndexException("Executor " + this.getClass().getSimpleName() + " is not compatible with bucket index");
```
NotSupportedException would be better?

minihippo replied:
done

leesf commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/utils/HiveHasher.java:
```diff
@@ -0,0 +1,362 @@
+
+import sun.misc.Unsafe;
+
+public class HiveHasher {
```
Please rename too.

minihippo replied:
HiveHasher means the unique hash function Hive used, differently from murmurHash

leesf commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/utils/HiveHasher.java:
```diff
@@ -0,0 +1,362 @@
+
+import sun.misc.Unsafe;
+
+public class HiveHasher {
```
Adding tests for the Class?

minihippo replied:
HiveHasher means the unique hash function is hive used, differently from `murmurHash`.

minihippo replied:
BucketUtilsTest already covers this part

leesf commented on hudi-client/hudi-client-common/src/test/java/org/apache/hudi/utils/TestHiveBucketUtils.java:
```diff
@@ -0,0 +1,62 @@
+
+import org.junit.jupiter.api.Test;
+
+public class TestHiveBucketUtils {
```
rename

minihippo replied:
done

leesf replied:
I did not see the updates you pushed.

leesf commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/index/bucket/SparkHiveBucketIndex.java:
```diff
@@ -0,0 +1,179 @@
+        .getProperty(HoodieIndexConfig.BUCKET_INDEX_BUCKET_NUM.key());
+    if (tableNumBucket == null && indexNumBucket == null) {
+      throw new IllegalArgumentException(
+          "Please set hoodie.index.hive.bucket.num or hoodie.table.numbuckets(deprecated) to a positive integer.");
```
hoodie.table.numbuckets is not exist? and please change `hoodie.index.hive.bucket.num` hardcoded to config.

minihippo replied:
`hoodie.table.numbuckets` is in the table config. Bucket info stored as a part of table meta data is beneficial to optimizing the query plan. The value of this config has to be the same with `HoodieIndexConfig.HIVE_BUCKET_INDEX_BUCKET_NUM`

minihippo replied:
Besides, bucket num of a table is always a fixed value and rarely change

minihippo replied:
I prefer to keep the `hoodie.table.numbuckets` instead of `hoodie.index.hive.bucket.num`

minihippo replied:
At last, `hoodie.index.hive.bucket.num` is used to get the table bucket num instead of 
 being stored in hoodie.properties. It is hard to access. Hive metastore server can be used to store the table bucket info.

leesf commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/index/bucket/SparkHiveBucketIndex.java:
```diff
@@ -0,0 +1,179 @@
+          "Please set hoodie.index.hive.bucket.num or hoodie.table.numbuckets(deprecated) to a positive integer.");
+    }
+    this.numBuckets = Integer.parseInt(indexNumBucket != null ? indexNumBucket : tableNumBucket);
+    LOG.info("use hive bucket index, numBuckets={}", numBuckets);
```
hive -> hudi

minihippo replied:
done

leesf commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/index/bucket/SparkHiveBucketIndex.java:
```diff
@@ -0,0 +1,179 @@
+
+  @Override
+  public boolean canIndexLogFiles() {
+    // for hive bucket index, it's not safe to send inserts(new data in one file group) to log files
```
rename hive

minihippo replied:
done

leesf commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/action/commit/SparkHiveBucketPartitioner.java:
```diff
@@ -0,0 +1,134 @@
+ *
+ * @param <T>
+ */
+public class SparkHiveBucketPartitioner<T extends HoodieRecordPayload<T>> extends
```
rename

minihippo replied:
done

leesf commented on hudi-common/src/main/java/org/apache/hudi/common/table/HoodieTableConfig.java:
```diff
@@ -140,6 +140,12 @@
 
   public static final String NO_OP_BOOTSTRAP_INDEX_CLASS = NoOpBootstrapIndex.class.getName();
 
+  public static final ConfigProperty<String> HOODIE_TABLE_NUM_BUCKETS = ConfigProperty
```
why not reuse `hoodie.bucket.index.bucket.num`?

minihippo replied:
done

leesf commented on hudi-common/src/main/java/org/apache/hudi/common/table/log/block/HoodieLogBlock.java:
```diff
@@ -45,8 +45,9 @@
+   * BD version starts from 1000 to avoid conflict wth community.
    */
-  public static int version = 1;
+  public static int version = 1000;
```
why need this change?

minihippo replied:
Modify to version 2

leesf commented on hudi-common/src/main/java/org/apache/hudi/common/util/serializer/HoodieKeyV1Serializer.java:
```diff
@@ -0,0 +1,45 @@
+/**
+ * Serializer for HoodieKey V1.
+ */
+public class HoodieKeyV1Serializer extends Serializer {
```
should we also introduce HoodieKeyV2Serializer?

minihippo replied:
done

minihippo replied:
Introduce the `HoodieKeyV${LatestVersion}Serializer` and register the serializer class according to log version by reflecting？

leesf commented:
@minihippo any updates?

minihippo commented:
Hi @leesf, i wil fix the unresolved comments in these two days

minihippo commented:
Hi @vinothchandar, can you take a look?

vinothchandar commented:
@minihippo Will do. Apologize if I am slow in the next few days. trying to push out the 0.9.0 RC

SoldierZh commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieWriteConfig.java:
```diff
@@ -933,6 +933,10 @@ public boolean getGlobalSimpleIndexUpdatePartitionPath() {
   }
 
+  public int getNumBuckets() {
+    return Integer.parseInt(props.getProperty(HoodieIndexConfig.BUCKET_INDEX_BUCKET_NUM.key(), "-1"));
```
Can we use the `BUCKET_INDEX_BUCKET_NUM.defaultValue()` here

minihippo replied:
fix

SoldierZh commented on hudi-common/src/main/java/org/apache/hudi/keygen/BaseKeyGenerator.java:
```diff
@@ -71,11 +75,28 @@ public final HoodieKey getKey(GenericRecord record) {
+  protected void validateIndexKeyField() {
+    if (indexKeyFields.size() != 0) {
+      for (String indexKeyField : indexKeyFields) {
+        if (!recordKeyFields.contains(indexKeyField)) {
```
Can try this way `recordKeyFields.containsAll(indexKeyFields)`

minihippo replied:
fix

minihippo commented:
HI @vinothchandar , is there any comments?

leesf commented:
@minihippo would you please address the comments above?

minihippo commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/keygen/ComplexAvroKeyGenerator.java:
```diff
@@ -47,4 +52,9 @@ public String getRecordKey(GenericRecord record) {
+
+  @Override
+  public List<Object> getIndexKey(GenericRecord record) {
+    return KeyGenUtils.getIndexKey(record, getIndexKeyFields());
```
similar to `getRecordKey`

minihippo commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/action/commit/SparkBucketIndexPartitioner.java:
```diff
@@ -0,0 +1,126 @@
+  private final int totalPartitionPaths;
+  private final List<String> partitionPaths;
+  private final Map<String, Integer> partitionPathOffset;
+  private Map<String, Set<String>> partitionPathFileIds;
```
good suggestion

minihippo commented:
> @minihippo would you please address the comments above?

done

leesf commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieIndexConfig.java:
```diff
@@ -200,6 +203,19 @@
+  // ***** Bucket Index Configs *****
+  public static final ConfigProperty<Integer> BUCKET_INDEX_BUCKET_NUM = ConfigProperty
+      .key("hoodie.index.bucket.num")
+      .defaultValue(-1)
```
@minihippo here why using -1 as the default value? can we use a recommended value such as 128/256 or other value for BUCKET index? because I think users may not know how to set the value.

vinothchandar replied:
+1

leesf commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/execution/CopyOnWriteInsertHandler.java:
```diff
@@ -85,7 +85,9 @@ public void consumeOneRecord(HoodieInsertValueGenResult<HoodieRecord> payload) {
-    if (!handle.canWrite(payload.record)) {
+    // If index requires a customized partitioner, cannot open a new file here since it might violate
+    // the distribution required by index.
+    if (!handle.canWrite(payload.record) && config.getNumBuckets() <= 0) {
```
I am confused about the `config.getNumBuckets <= 0`, it means requires a customized partitioner?

minihippo replied:
yes, add a new api `HoodieIndex.needCustomizedPartitioner` to do the thing like this

minihippo commented:
> @minihippo Will do. Apologize if I am slow in the next few days. trying to push out the 0.9.0 RC

Hi @vinothchandar , do you have any time to review this pr?

vinothchandar commented:
@minihippo Working on it! Really sorry for the long delay, the 0.9.0 release was lot more work than I originally budgeted time for. This is a needle mover for the next release!

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieIndexConfig.java:
```diff
@@ -200,6 +203,19 @@
+      .withDocumentation("Only applies if index type is BUCKET_INDEX. Determine the bucket num of the hudi table, "
+          + "and each partition is divided to N buckets.");
+
+  public static final ConfigProperty<String> INDEX_KEY_FILED_NAME = ConfigProperty
```
lets name everything consistently using `BUCKET_INDEX_*` prefix? Also the typo `FILED` -> `FIELD`

minihippo replied:
done

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieIndexConfig.java:
```diff
@@ -200,6 +203,19 @@
+  public static final ConfigProperty<Integer> BUCKET_INDEX_BUCKET_NUM = ConfigProperty
+      .key("hoodie.index.bucket.num")
+      .defaultValue(-1)
+      .withDocumentation("Only applies if index type is BUCKET_INDEX. Determine the bucket num of the hudi table, "
```
Just. `Determine the total number of buckets in each table partition.` ?

minihippo replied:
I has already implemented partition-level bucket, but not included in this pr. Besides, the partition-level bucket optimization rules is introduced to Spark for efficient reading. Both of them will be contributed to community in the following prs

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieIndexConfig.java:
```diff
@@ -200,6 +203,19 @@
 
+  // ***** Bucket Index Configs *****
+  public static final ConfigProperty<Integer> BUCKET_INDEX_BUCKET_NUM = ConfigProperty
+      .key("hoodie.index.bucket.num")
```
`hoodie.bucket.index.num.buckets`?

minihippo replied:
done

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieIndexConfig.java:
```diff
@@ -200,6 +203,19 @@
       .withDocumentation("Similar to " + BLOOM_INDEX_UPDATE_PARTITION_PATH_ENABLE + ", but for simple index.");
 
+  // ***** Bucket Index Configs *****
+  public static final ConfigProperty<Integer> BUCKET_INDEX_BUCKET_NUM = ConfigProperty
```
`BUCKET_INDEX_NUM_BUCKETS`?

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieIndexConfig.java:
```diff
@@ -200,6 +203,19 @@
+      .key("hoodie.index.key.field")
+      .noDefaultValue()
+      .withAlternatives("hoodie.datasource.write.indexkey.field")
+      .withDocumentation("Index key.It is used to index the record and find its file group");
```
this is the field we hash by? if so, call this `BUCKET_INDEX_HASH_FIELD` ?  and also fix the property name?

minihippo replied:
done

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieWriteConfig.java:
```diff
@@ -1372,6 +1372,10 @@ public boolean getGlobalSimpleIndexUpdatePartitionPath() {
     return getBoolean(HoodieIndexConfig.SIMPLE_INDEX_UPDATE_PARTITION_PATH_ENABLE);
   }
 
+  public int getNumBuckets() {
```
lets make this more descriptive. `getBucketIndexNumBuckets()`?

minihippo replied:
done

vinothchandar reviewed (commented):
Just getting started

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/keygen/ComplexAvroKeyGenerator.java:
```diff
@@ -36,6 +38,9 @@ public ComplexAvroKeyGenerator(TypedProperties props) {
         .split(",")).map(String::trim).filter(s -> !s.isEmpty()).collect(Collectors.toList());
     this.partitionPathFields = Arrays.stream(props.getString(KeyGeneratorOptions.PARTITIONPATH_FIELD_NAME.key())
         .split(",")).map(String::trim).filter(s -> !s.isEmpty()).collect(Collectors.toList());
+    this.indexKeyFields = props.getStringList(
```
Also why would this be different than the record key fields? We bucket by those right, only then Can we make sure Updates to same key k goes into the same bucket?

minihippo replied:
> Can we keep the key generator unchanged? Extract the index field during actual
> indexing? I find it odd that key generator has to know about index field. I guess you
> want to do this for efficiency?

Keep the index field is to be more efficient without deserialization

minihippo replied:
Bucket key needs to be a subset of the record key,  so that records with the same key can go into the same bucket. e.g. bucket key is `colA` and record key is `id` + `colA`.
Generally, the record key is `id` to identified the record, and bucket key is more like an attribute to cluster the data. When using bucket index, bucket key needs to be added to record key to meet the data distribution and uniquely identify a record.

minihippo replied:
Good suggestion, will fix

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/keygen/KeyGenUtils.java:
```diff
@@ -135,6 +136,20 @@ public static String getRecordPartitionPath(GenericRecord record, List<String> p
+   */
+  public static List<Object> getIndexKey(GenericRecord record, List<String> indexKeyFields) {
+    if (indexKeyFields == null || indexKeyFields.size() == 0) {
+      return null;
```
Let's please not return nulls. Option?

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/commit/BaseCommitActionExecutor.java:
```diff
@@ -76,6 +77,11 @@ public BaseCommitActionExecutor(HoodieEngineContext context, HoodieWriteConfig c
     // TODO : Remove this once we refactor and move out autoCommit method from here, since the TxnManager is held in {@link AbstractHoodieWriteClient}.
     this.txnManager = new TransactionManager(config, table.getMetaClient().getFs());
     this.lastCompletedTxn = TransactionUtils.getLastCompletedTxnInstantAndMetadata(table.getMetaClient());
+    // TODO: HUDI-2155 bulk insert support bucket index, HUDI-2156 cluster the table with bucket index.
```
Can we tackle it in the same PR? bulk. insert is pretty foundational

minihippo replied:
Agree with u, bulk insert is foundational and it has been already implemented. But can we merge this pr first? I consider the bulk insert part will make this pr time period more longer.

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/commit/BaseCommitActionExecutor.java:
```diff
@@ -76,6 +77,11 @@ public BaseCommitActionExecutor(HoodieEngineContext context, HoodieWriteConfig c
     this.lastCompletedTxn = TransactionUtils.getLastCompletedTxnInstantAndMetadata(table.getMetaClient());
+    // TODO: HUDI-2155 bulk insert support bucket index, HUDI-2156 cluster the table with bucket index.
+    if (this.config.getIndexType() == HoodieIndex.IndexType.BUCKET_INDEX
+        && (operationType == WriteOperationType.BULK_INSERT || operationType == WriteOperationType.CLUSTER)) {
```
We might add new write operations and this check would let them through. e-g- Optimize introduced by
z-order PR

vinothchandar replied:
Can we rewrite this based on what is actually Supported On bucket index

minihippo replied:
will fix

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/utils/HiveHasher.java:
```diff
@@ -0,0 +1,196 @@
+
+import sun.misc.Unsafe;
+
+public class HiveHasher {
```
Why do we specifically need the hashused by hive?

vinothchandar replied:
This will be Something we will have to now maintain for a long time. I Would prefer we used Some simpler performant hashing aIgorithm that came with the Jvm

garyli1019 replied:
or make the hash method an interface and configurable? Flink uses murmurhash. We might extend this for different use cases later.

vinothchandar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/index/bucket/SparkBucketIndex.java:
```diff
@@ -0,0 +1,183 @@
+import org.apache.hudi.table.WorkloadProfile;
+import org.apache.hudi.table.action.commit.Partitioner;
+import org.apache.hudi.table.action.commit.SparkBucketIndexPartitioner;
+import org.jetbrains.annotations.NotNull;
```
jetbrains? can weavoid this

vinothchandar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/index/bucket/SparkBucketIndex.java:
```diff
@@ -0,0 +1,183 @@
+import org.apache.hudi.table.action.commit.SparkBucketIndexPartitioner;
+import org.jetbrains.annotations.NotNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
```
log4j?

vinothchandar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/index/bucket/SparkBucketIndex.java:
```diff
@@ -0,0 +1,183 @@
+      HoodieEngineContext context,
+      HoodieTable<T, JavaRDD<HoodieRecord<T>>, JavaRDD<HoodieKey>, JavaRDD<WriteStatus>> hoodieTable)
+      throws HoodieIndexException {
+    List<String> partitions = records.map(HoodieRecord::getPartitionPath)
```
I think our goal here has to be to do this in a Streaming fashion Without any sort
off caching. Even as it stands now, we will double
read from source due to spark recompute since we are not Caching any input RDD

vinothchandar reviewed (changes requested):
Still reviewing

leesf commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/index/HoodieIndex.java:
```diff
@@ -97,13 +100,32 @@ public abstract O updateLocation(O writeStatuses, HoodieEngineContext context,
+   * If the `getCustomizedPartitioner` returns a partitioner, it has to be true.
+   */
+  @PublicAPIMethod(maturity = ApiMaturityLevel.EVOLVING)
+  public boolean needCustomizedPartitioner() {
```
can we remove this method and replace the callers of `needCustomzedPartitioner` to `getCustomzedPartitioner`, and determine whether it returns Option.empty?

minihippo replied:
Work profile cannot be accessed where we call the `needCustomizedPartitioner `

leesf commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/index/bucket/SparkBucketIndex.java:
```diff
@@ -0,0 +1,183 @@
+
+  public SparkBucketIndex(HoodieWriteConfig config) {
+    super(config);
+    String indexNumBucket = config.getProps()
```
we would move getting `HoodieIndexConfig.BUCKET_INDEX_BUCKET_NUM.key()` into HoodieWriteConfig?

minihippo replied:
Done. I am confused about getting index config from the writing config. In my opinion, index type and the other config are determined at the first time. It is not user-friendly that set the same and lengthy index related kvs every time. it would be better to store the index config in hoodie.properties?

leesf commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/index/bucket/SparkBucketIndex.java:
```diff
@@ -0,0 +1,183 @@
+      throw new IllegalArgumentException(
+          "Please set hoodie.index.hive.bucket.num or hoodie.table.numbuckets(deprecated) to a positive integer.");
+    }
+    this.numBuckets = Integer.parseInt(indexNumBucket);
```
moving to `HoodieWriteConfig` would avoid the this.

minihippo commented:
Hi all, i will fix all comments in this week

garyli1019 commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/utils/BucketUtils.java:
```diff
@@ -0,0 +1,89 @@
+  }
+
+  public static String newBucketFileIdPrefix(String bucketId) {
+    return FSUtils.createNewFileIdPfx().replaceFirst(".{8}", bucketId);
```
wondering if we still need the full fileId or we can just use the bucketId as the fileId, cause it's already unique under the partition.

nsivabalan commented:
I am going to review this and help you take this across the finish line. give me a day or two.

nsivabalan commented:
@minihippo : While I am starting to review, curious to know if you get a chance to do any perf analysis with existing bloom index.

vinothchandar commented:
@minihippo sorry not mention here. @vingov is also going to try and test this as well for uber

nsivabalan reviewed (changes requested):
Thanks for the contribution. Have left some comments. 
And IIUC, not all clustering strategies may sit well with this bucket index. Some clustering are intended to create new file groups which may not work out in case of bucket index right? 
We should call this out somehwere as to what are the constraints with this type of index. like every growing file sizes, no small file handling, etc.

nsivabalan commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/action/commit/SparkBucketIndexPartitioner.java:
```diff
@@ -0,0 +1,128 @@
+  }
+
+  @Override
+  public BucketInfo getBucketInfo(int bucketNumber) {
```
just so we are on same page or to confirm my understanding. This getBucketInfo(bucketNum) should be called only once per bucket for a non-existant bucket. If not, 
```
BucketInfo(BucketType.INSERT, BucketUtils.newBucketFileIdPrefix(bucketId), partitionPath);
```
this line, could create diff fileId for N number of calls for same bucket. since we prefix some random value to fileId apart from bucket Id, this is not idempotent for a given bucket. 
Just wanted to call this out. 

Let me know if my statement is not clear. I can give more context.

minihippo replied:
U are right, bucketId 0 in partition p1 and partition p2 are the different file groups. The first 8 digits of the UUID equals bucketId, followed by random numbers.

minihippo commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieIndexConfig.java:
```diff
@@ -200,6 +202,25 @@
+
+  public static final ConfigProperty<String> BUCKET_INDEX_HASH_FIELD = ConfigProperty
+      .key("hoodie.bucket.index.hash.field")
+      .noDefaultValue()
```
Good suggestion. Do this in validating when build the index config

minihippo commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieIndexConfig.java:
```diff
@@ -200,6 +202,25 @@
+      .withDocumentation("Index key. It is used to index the record and find its file group");
+
+  public static final ConfigProperty<String> BUCKET_INDEX_HASH_FUNCTION = ConfigProperty
+      .key("hoodie.bucket.index.hash.function")
```
HashFunction is a simple interface to achieve flexible extension of hash function. Currently we support JVM hash and Hive hash

minihippo commented on hudi-common/src/main/java/org/apache/hudi/common/model/HoodieKey.java:
```diff
@@ -31,13 +35,16 @@
   }
 
-  public HoodieKey(String recordKey, String partitionPath) {
+  public HoodieKey(String recordKey, String partitionPath, List<Object> indexKey) {
```
Index key is used to tag location and cluster the insert/upsert data. To avoid repeated serialization and deserialization, can we store the index key and save it in `HoodieRecord`?

minihippo commented:
> @minihippo : While I am starting to review, curious to know if you get a chance to do any perf analysis with existing bloom index.

In our case, an unpartitioned table with 40000 file group, over 30TB and more than 500 billion records cannot run successfully with bloom index. Besides, hudi is 0.6.0.

minihippo commented:
> Thanks for the contribution. Have left some comments. And IIUC, not all clustering strategies may sit well with this bucket index. Some clustering are intended to create new file groups which may not work out in case of bucket index right? We should call this out somehwere as to what are the constraints with this type of index. like every growing file sizes, no small file handling, etc.

Yes, you are right. Should I add a doc to the index config to explain the constraints?

minihippo commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/index/bucket/SparkBucketIndex.java:
```diff
@@ -0,0 +1,152 @@
+      if (!partitionPathFileIDList.containsKey(partitionPath)) {
+        partitionPathFileIDList.put(partitionPath, loadPartitionBucketIdFileIdMapping(hoodieTable, partitionPath));
+      }
+      if (partitionPathFileIDList.get(partitionPath).containsKey(bucketId)) {
```
will fix

minihippo commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/action/commit/SparkBucketIndexPartitioner.java:
```diff
@@ -0,0 +1,128 @@
+  public int getPartition(Object key) {
+    Tuple2<HoodieKey, Option<HoodieRecordLocation>> keyLocation =
+        (Tuple2<HoodieKey, Option<HoodieRecordLocation>>) key;
+    String partitionPath = keyLocation._1.getPartitionPath();
```
will fix

minihippo commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/action/commit/SparkHoodiePartitioner.java:
```diff
@@ -0,0 +1,48 @@
+import org.apache.hudi.table.WorkloadProfile;
+import org.apache.spark.Partitioner;
+
+public abstract class SparkHoodiePartitioner<T extends HoodieRecordPayload<T>> extends Partitioner
```
will fix

minihippo commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/action/deltacommit/AbstractSparkDeltaCommitActionExecutor.java:
```diff
@@ -74,23 +74,25 @@ public Partitioner getUpsertPartitioner(WorkloadProfile profile) {
       LOG.info("Small file corrections for updates for commit " + instantTime + " for file " + fileId);
       return super.handleUpdate(partitionPath, fileId, recordItr);
-    } else {
+    } else if (recordItr.hasNext()) {
```
will fix

minihippo commented on hudi-common/src/main/java/org/apache/hudi/common/util/serializer/HoodieKeyV1Serializer.java:
```diff
@@ -0,0 +1,45 @@
+  }
+
+  @Override
+  public Object read(Kryo kryo, Input input, Class type) {
```
This change is introduced by adding index key to hoodie key. Will fix

minihippo commented on hudi-common/src/main/java/org/apache/hudi/common/util/SerializationUtils.java:
```diff
@@ -78,15 +83,29 @@
+    if (objectData == null) {
+      throw new IllegalArgumentException("The byte[] must not be null");
+    }
+    if (version == HoodieLogBlock.version) {
```
This change is introduced by adding index key to hoodie key. Will fix

minihippo commented on hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/hudi/MergeOnReadSnapshotRelation.scala:
```diff
@@ -146,7 +154,7 @@ class MergeOnReadSnapshotRelation(val sqlContext: SQLContext,
   }
 
-  def buildFileIndex(filters: Array[Filter]): List[HoodieMergeOnReadFileSplit] = {
+  def buildFileIndex(filters: Array[Filter]): List[List[HoodieMergeOnReadFileSplit]] = {
```
> may I know why this change?

For Spark, it can treat the hudi table with bucket index as a bucketed table and then each task will read the files  belonging to the same bucket. If reading only one partition, a task will only read a file group. But if more than one, to bucket table, each task will read the files which has the same bucket id and from different partition. Thus each task  will have more one file split.
Treating as a bucketed table will accelerated query, especially to the join/group by bucket index key query.

minihippo commented on hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/hudi/HoodieMergeOnReadRDD.scala:
```diff
@@ -58,32 +64,50 @@ class HoodieMergeOnReadRDD(@transient sc: SparkContext,
-        s"spark partition Index: ${mergeOnReadPartition.index}" +
-        s"merge type: ${mergeOnReadPartition.split.mergeType}")
-    }
+    mergeOnReadPartition.split.iterator.flatMap { sp =>
```
Did u mean that reading optimization for hudi table with bucket index needs to have another pr?

minihippo commented on hudi-common/src/main/java/org/apache/hudi/keygen/BaseKeyGenerator.java:
```diff
@@ -71,11 +75,24 @@ public final HoodieKey getKey(GenericRecord record) {
+  /**
+   * If index key exists, it should be the subset of the record key list.
+   */
+  protected void validateIndexKeyField() {
```
No, only field in the record field can guarantee value's immutability. Once the record value changes, the updated record cannot find the file group where it was originally located.

minihippo commented:
Hi @nsivabalan，thanks for review. According to the comments, one doubtful is where the index key is placed, deserialize each time if using, or save it once deserialize?

minihippo commented:
Hi @nsivabalan, I've fixed all comments. The main changes are:
1. Unify bucket index configurations to the HoodieIndexConfig
2. On the premise that bucket index key has to be the subset of the record key, get the index key value at the runtime from HoodieKey without destroy the data structure. `BucketIdentifier` is introduced to do it.
3. When `tag location`, cache the partial filesystem view in each Spark task. The implementation is different from bloom index which cache hoodie key and file name first and then join the input data. Bucket Index is proposed to processing more bigger data and join is a heavy operation. Therefore, hoodieRecordRDD to taggedRecordRDD is a mapPartition only operation.

leesf commented:
@nsivabalan can we make this a release-blocker and make it to the next release since it has been a long time since opened, and @minihippo would resolve the comments

minihippo commented:
@hudi-bot run azure

minihippo commented:
> Hi @nsivabalan, I've fixed all comments. The main changes are:
> 
> 1. Unify bucket index configurations to the HoodieIndexConfig
> 2. On the premise that bucket index key has to be the subset of the record key, get the index key value at the runtime from HoodieKey by a tricky way without destroying the data structure. `BucketIdentifier` is introduced to do it.
> 3. When `tag location`, cache the partial filesystem view in each Spark task. The implementation is different from bloom index which caches hoodie key and file name first and then join with the input data. Bucket Index is proposed to processing more bigger data and join is a heavy operation. Therefore, hoodieRecordRDD to taggedRecordRDD is a mapPartition only operation.

friendly ping @nsivabalan,  could you please take a look in your spare time?

vinothchandar commented:
@minihippo Wondering where we are on this. We can get this in to 0.10 if the changes are mostly isolated. let me know.

minihippo commented:
> @minihippo Wondering where we are on this. We can get this in to 0.10 if the changes are mostly isolated. let me know.

@vinothchandar thanks for replying. Does `isolated` mean that bucket index will not affect basic functions and other features? Whether it is used is controlled by the switch.
Currently, i addressed all comments added by @nsivabalan, but i'm not sure the changes are acceptable to the community.

minihippo commented:
> Hi @nsivabalan, I've fixed all comments. The main changes are:
> 
> 1. Unify bucket index configurations to the HoodieIndexConfig
> 2. On the premise that bucket index key has to be the subset of the record key, get the index key value at the runtime from HoodieKey by a tricky way without destroying the data structure. `BucketIdentifier` is introduced to do it.
> 3. When `tag location`, cache the partial filesystem view in each Spark task. The implementation is different from bloom index which caches hoodie key and file name first and then join with the input data. Bucket Index is proposed to processing more bigger data and join is a heavy operation. Therefore, hoodieRecordRDD to taggedRecordRDD is a mapPartition only operation.

@vinothchandar, here is the summary after all comments addressed

YuweiXiao commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieIndexConfig.java:
```diff
@@ -200,6 +209,48 @@
+          + "Possible options are [JVMHash | HiveHash]. ");
+
+  public static final Set<WriteOperationType> BUCKET_INDEX_SUPPORTED_OPERATIONS = new HashSet<WriteOperationType>() {{
+      add(WriteOperationType.INSERT);
```
The current implementation seems not support insert. Two consecutive insertions (with records going into same bucket) will produce two different parquet files with the same bucketId.

minihippo replied:
It is caused by insert operation doesn't tag location before creating the work profile.  Will fix.

YuweiXiao commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieIndexConfig.java:
```diff
@@ -54,7 +63,7 @@
       .noDefaultValue()
       .withDocumentation("Type of index to use. Default is Bloom filter. "
-          + "Possible options are [BLOOM | GLOBAL_BLOOM |SIMPLE | GLOBAL_SIMPLE | INMEMORY | HBASE]. "
+          + "Possible options are [BLOOM | GLOBAL_BLOOM |SIMPLE | GLOBAL_SIMPLE | INMEMORY | HBASE | BUCKET]. "
```
BUCKET_INDEX rather than BUCKET, should be consistent with HoodieIndex::IndexType

minihippo replied:
will fix

vinothchandar commented:
@minihippo As the 0.10 voting goes on, lets work together to land this. I ll make sometime later this or over weekend. Thanks for your patience.

minihippo commented:
Progress Update：
- support consecutive insertions  cc @YuweiXiao @leesf

vinothchandar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/index/bucket/SparkBucketIndex.java:
```diff
@@ -0,0 +1,174 @@
+              numBuckets, config.getBucketIndexHashFunction());
+          String partitionPath = record.getPartitionPath();
+          if (!partitionPathFileIDList.containsKey(partitionPath)) {
+            partitionPathFileIDList.put(partitionPath, loadPartitionBucketIdFileIdMapping(hoodieTable, partitionPath));
```
rename: partitionPath

minihippo replied:
don't get the point, can u explain this?

vinothchandar commented on hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/hudi/HoodieMergeOnReadRDD.scala:
```diff
@@ -39,7 +39,13 @@ import scala.collection.JavaConverters._
 
-case class HoodieMergeOnReadPartition(index: Int, split: HoodieMergeOnReadFileSplit) extends Partition
+/**
+ * Files a input task reads. Usually a file spilt corresponds to a task. For partitioned hudi table with bucket index,
```
IIUC we just replace the first 8 of the fileID prefix right. so even though the bucket numbers are same, the file names will be different/unique across table?

minihippo replied:
Yes. files with same bucketid but under the different partitions have different filenames

vinothchandar commented on hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/hudi/MergeOnReadSnapshotRelation.scala:
```diff
@@ -219,17 +226,31 @@ class MergeOnReadSnapshotRelation(val sqlContext: SQLContext,
     }
   }
+
+  private def groupFilesWithBucket(
```
can you explain why this needs to happen? At a high level, I would expect bucketing not alter how we group files? we have 1 file group per bucket right?

minihippo replied:
This is a read optimizer. I revert this changes and will create another pr to cover the query optimizer part.

vinothchandar reviewed (commented):
Made a full pass. Some cleanups to do (some  in the PR, some we can defer). Also few clarifying questions. This is looking in good shape actually. So we should be able to iterate quickly. Thanks again for your patience!

minihippo commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/index/HoodieIndex.java:
```diff
@@ -122,13 +125,40 @@ public O updateLocation(O writeStatuses, HoodieEngineContext context,
+  /**
+   * If the `getCustomizedPartitioner` returns a partitioner, it has to be true.
+   */
+  @PublicAPIMethod(maturity = ApiMaturityLevel.EVOLVING)
```
already fix

minihippo commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/action/commit/BaseSparkCommitActionExecutor.java:
```diff
@@ -214,7 +214,11 @@ private void initKeyGenIfNeeded(boolean populateMetaFields) {
 
   protected Partitioner getPartitioner(WorkloadProfile profile) {
-    if (WriteOperationType.isChangingRecords(operationType)) {
+    Option<org.apache.hudi.table.action.commit.Partitioner> customizedPartitioner = table.getIndex()
```
already fix

minihippo commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/action/deltacommit/AbstractSparkDeltaCommitActionExecutor.java:
```diff
@@ -74,8 +74,8 @@ public Partitioner getUpsertPartitioner(WorkloadProfile profile) {
     LOG.info("Merging updates for commit " + instantTime + " for file " + fileId);
-
-    if (!table.getIndex().canIndexLogFiles() && mergeOnReadUpsertPartitioner.getSmallFileIds().contains(fileId)) {
+    if (!table.getIndex().canIndexLogFiles() && mergeOnReadUpsertPartitioner != null
```
mergeOnReadUpsertPartitioner cannot be initialized when using Bucketpartitioner.
The `AbstractSparkDeltaCommitActionExcutor` holding the partitioner is to rewrite the small parquet when update. Getting small files from the partitioner is strange here. IMO, small files should be maintained by filesystemview. Shall i create another pr to resolve this?

vinothchandar replied:
let me take a closer look and suggest what I think

minihippo commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieIndexConfig.java:
```diff
@@ -200,6 +209,48 @@
+      add(WriteOperationType.COMPACT);
+      add(WriteOperationType.DELETE_PARTITION);
+      // TODO: HUDI-2155 bulk insert support bucket index.
+      // TODO: HUDI-2156 cluster the table with bucket index.
```
yes

minihippo commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieIndexConfig.java:
```diff
@@ -540,5 +602,28 @@ private String getDefaultIndexType(EngineType engineType) {
+      if (hoodieIndexConfig.getString(INDEX_TYPE)
+          .equalsIgnoreCase(HoodieIndex.IndexType.BUCKET_INDEX.toString())) {
+        // check the bucket index hash field
+        if (StringUtils.isNullOrEmpty(hoodieIndexConfig.getString(BUCKET_INDEX_HASH_FIELD))) {
```
I check that we set KeyGeneratorOptions.RECORDKEY_FIELD_NAME  across all code paths

minihippo commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieIndexConfig.java:
```diff
@@ -200,6 +209,48 @@
+  /**
+   * ***** Bucket Index Configs *****
+   * Bucket Index is targeted to locate the record fast by hash in big data scenarios.
+   * The current implementation is a basic version, so there are some constraints:
```
HUDI-3039

minihippo commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieIndexConfig.java:
```diff
@@ -200,6 +209,48 @@
+      .withDocumentation("Hash function. It is used to compute the index key hash value "
+          + "Possible options are [JVMHash | HiveHash]. ");
+
+  public static final Set<WriteOperationType> BUCKET_INDEX_SUPPORTED_OPERATIONS = new HashSet<WriteOperationType>() {{
```
move to the storage layout

minihippo commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/execution/CopyOnWriteInsertHandler.java:
```diff
@@ -85,7 +85,9 @@ public void consumeOneRecord(HoodieInsertValueGenResult<HoodieRecord> payload) {
-    if (!handle.canWrite(payload.record)) {
+    // If index requires a customized partitioner, cannot open a new file here since it might violate
+    // the distribution required by index.
+    if (!handle.canWrite(payload.record) && !hoodieTable.getIndex().needCustomizedPartitioner()) {
```
> this is a little opaque. but get why you are doing this. I ll think of better ways if possible
----
Bucket limit that only one parquet file in a bucket. It;s a bucket layout limitation for writer, so i abstract the constraint and move it to the storage layout

minihippo replied:
> how do we handle this for MOR/AppendHandle?

No limitation. AppendHandle can write another log file with `log_ver_2`.

minihippo commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/commit/AbstractWriteHelper.java:
```diff
@@ -46,7 +46,7 @@
       Instant lookupBegin = Instant.now();
       I taggedRecords = dedupedRecords;
-      if (performTagging) {
+      if (performTagging || table.getIndex().needTaggingIfInsert()) {
```
will fix

minihippo commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/index/bucket/BucketIdentifier.java:
```diff
@@ -0,0 +1,84 @@
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+public class BucketIdentifier {
```
will fix

minihippo commented on hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/DataSourceOptions.scala:
```diff
@@ -526,6 +526,12 @@ object DataSourceWriteOptions {
     .noDefaultValue()
     .withDocumentation("Mode to choose for Hive ops. Valid values are hms, jdbc and hiveql.")
 
+  // bucketSpec: CLUSTERED BY (trace_id) SORTED BY (trace_id ASC) INTO 65536 BUCKETS
```
> add this to the doc? is this an example bucket spec?

user documents ? yes, hive style.

vinothchandar replied:
need to think if there are better ways of exposing this config to the user

minihippo replied:
already fix

minihippo commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/index/SparkHoodieIndexFactory.java:
```diff
@@ -63,6 +64,8 @@ public static HoodieIndex createIndex(HoodieWriteConfig config) {
         return new HoodieSimpleIndex<>(config, getKeyGeneratorForSimpleIndex(config));
       case GLOBAL_SIMPLE:
         return new HoodieGlobalSimpleIndex<>(config, getKeyGeneratorForSimpleIndex(config));
+      case BUCKET_INDEX:
```
will fix

minihippo commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/index/bucket/BucketIdentifier.java:
```diff
@@ -0,0 +1,84 @@
+          .map(f -> recordKeyPairs.get(f))
+          .collect(Collectors.toList());
+    }
+    return mod(HashFunction.getHashFunction(hashFunction).hash(hashKeyFields)
```
will fix

minihippo replied:
hashcode can be negative, & to let always bucketId >= 0

minihippo commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/index/bucket/BucketIdentifier.java:
```diff
@@ -0,0 +1,84 @@
+
+  private static int mod(int x, int y) {
+    int r = x % y;
+    if (r < 0) {
```
will fix

minihippo commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieIndexConfig.java:
```diff
@@ -200,6 +209,48 @@
+
+  public static final ConfigProperty<String> BUCKET_INDEX_HASH_FUNCTION = ConfigProperty
+      .key("hoodie.bucket.index.hash.function")
+      .defaultValue("JVMHash")
```
Revert the hashfunction, and just use object.hashcode to make it simple and efficient.

minihippo commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/index/bucket/BucketIdentifier.java:
```diff
@@ -0,0 +1,84 @@
+
+  public static int getBucketId(HoodieKey hoodieKey, String indexKeyFields, int numBuckets, String hashFunction) {
+    List<String> hashKeyFields;
+    if (!hoodieKey.getRecordKey().contains(":")) {
```
Hudi is commonly used to store data warehouse summary tables at ByteDance. Summary table is organized with dimensions and indexes. Id and dimension is composite primary key. Bucket key can be one of the dimension which is frequently aggregated by data analytics to accelerated query. In the above scenario, the primary key is a complex type.

minihippo commented:
The main changes are:
1. introduce the `layout` entry to constraint the write behavior
2. remove the abstraction of hash function, using jvm hashcode instead to make it simple
3. remove the changes about spark MergeOnReadRDD

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/commit/BaseCommitActionExecutor.java:
```diff
@@ -76,6 +76,10 @@ public BaseCommitActionExecutor(HoodieEngineContext context, HoodieWriteConfig c
     // TODO : Remove this once we refactor and move out autoCommit method from here, since the TxnManager is held in {@link AbstractHoodieWriteClient}.
     this.txnManager = new TransactionManager(config, table.getMetaClient().getFs());
     this.lastCompletedTxn = TransactionUtils.getLastCompletedTxnInstantAndMetadata(table.getMetaClient());
+    if (table.getStorageLayout().operationConstraint(operationType)) {
```
Thanks for doing this. the storage layout abstraction is something I wish we added before. but when done with real use-cases, makes it lot easier.

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/storage/HoodieStorageLayout.java:
```diff
@@ -0,0 +1,51 @@
+
+import java.io.Serializable;
+
+public class HoodieStorageLayout implements Serializable {
```
nts: Review these classes for names, docs

vinothchandar commented on hudi-client/hudi-client-common/src/test/resources/hive_bucket_id_check.csv:
```diff
@@ -0,0 +1,27 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+true,0,527706,0.9105978527125206,0.5821051620180587,57
```
I'd avoid checking in these data files

minihippo replied:
will fix

vinothchandar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/action/commit/BaseSparkCommitActionExecutor.java:
```diff
@@ -303,7 +305,7 @@ protected void commit(Option<Map<String, String>> extraMetadata, HoodieWriteMeta
   protected Iterator<List<WriteStatus>> handleUpsertPartition(String instantTime, Integer partition, Iterator recordItr,
                                                               Partitioner partitioner) {
-    UpsertPartitioner upsertPartitioner = (UpsertPartitioner) partitioner;
+    SparkHoodiePartitioner upsertPartitioner = (SparkHoodiePartitioner) partitioner;
```
is this renamed?

minihippo replied:
Introduce a `SparkHoodiePartitioner` to implements both spark partitioner and hudi partitioner which is to get bucket info. The `UpsertPartitioner` inherit it.

vinothchandar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/action/commit/SparkHoodiePartitioner.java:
```diff
@@ -0,0 +1,51 @@
+
+  protected final HoodieTable table;
+
+  public SparkHoodiePartitioner(WorkloadProfile profile,
```
wrapping

vinothchandar commented on hudi-client/hudi-spark-client/src/test/java/org/apache/hudi/index/bucket/BucketIdentifierSuite.java:
```diff
@@ -0,0 +1,65 @@
+import org.apache.hudi.testutils.KeyGeneratorTestUtilities;
+import org.junit.jupiter.api.Test;
+
+public class BucketIdentifierSuite {
```
renaming consistent with other test classes

minihippo replied:
will fix

vinothchandar commented on hudi-client/hudi-spark-client/src/test/java/org/apache/hudi/index/bucket/HoodieBucketIndexSuite.java:
```diff
@@ -0,0 +1,147 @@
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class HoodieBucketIndexSuite extends HoodieClientTestHarness {
```
rename

minihippo replied:
will fix

vinothchandar commented on hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/functional/TestMORDataSourceWithBucket.scala:
```diff
@@ -0,0 +1,153 @@
+/**
+ *
+ */
+class TestDataSourceForBucketIndex extends HoodieClientTestBase {
```
Could we parameterize an existung test>?

minihippo replied:
the test case like `testCount` can be parameterized, but the `testDoubleInsert` is more like a bucket index unique test case?

minihippo replied:
https://issues.apache.org/jira/browse/HUDI-3121, and link to bucket index jira https://issues.apache.org/jira/projects/HUDI/issues/HUDI-3039

vinothchandar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/action/commit/SparkBucketIndexPartitioner.java:
```diff
@@ -0,0 +1,126 @@
+    assignUpdates(profile);
+  }
+
+  private void assignUpdates(WorkloadProfile profile) {
```
see if we can rewrite these functional style?

minihippo replied:
Can u explain it? doesn't understand

vinothchandar replied:
using java streams instead of for loops. its a minor comment

vinothchandar commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/action/commit/SparkBucketIndexPartitioner.java:
```diff
@@ -0,0 +1,126 @@
+  private final int totalPartitionPaths;
+  private final List<String> partitionPaths;
+  private final Map<String, Integer> partitionPathOffset;
+  private Map<String, Set<String>> updatePartitionPathFileIds;
```
docs. for these.

minihippo replied:
will fix

vinothchandar reviewed (approved):
LGTM, nearing landing. I will push some cleanups/changes in a day.

vinothchandar reviewed (approved):
@minihippo Thanks for the push, I also cleaned up some docs, naming. Please take a look at the last commit. Once CI passes and you are happy with it, I'll land. I have pointed out some follow-ups in this review. If you could add subtasks for them and track them to completion, that would be awesome.

vinothchandar commented:
There are some test failures 

```
R] Errors: 
[ERROR]   TestHoodieDeltaStreamer.testBulkInsertsAndUpsertsWithSQLBasedTransformerFor2StepPipeline:1088 » IllegalArgument
[ERROR]   TestHoodieDeltaStreamer.testPayloadClassUpdate:1176 » IllegalArgument Property...
[ERROR]   TestHoodieDeltaStreamer.testPayloadClassUpdateWithCOWTable:1202 » IllegalArgument
[INFO] 

[ERROR] Errors: 
[ERROR]   TestDataSourceUtils.testBuildHiveSyncConfig:224 » IllegalArgument Property hoo...
[ERROR]   TestDataSourceUtils.testBuildHiveSyncConfig:224 » IllegalArgument Property hoo...

```

Don't think these are related to my pushes. @minihippo Could you please check

minihippo commented on hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/DataSourceOptions.scala:
```diff
@@ -526,6 +526,12 @@ object DataSourceWriteOptions {
     .noDefaultValue()
     .withDocumentation("Mode to choose for Hive ops. Valid values are hms, jdbc and hiveql.")
 
+  val HIVE_SYNC_BUCKET_SYNC: ConfigProperty[Boolean] = ConfigProperty
```
done

minihippo commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/storage/HoodieStorageLayout.java:
```diff
@@ -0,0 +1,51 @@
+
+import java.io.Serializable;
+
+public class HoodieStorageLayout implements Serializable {
```
fix

minihippo commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/storage/HoodieBucketLayout.java:
```diff
@@ -0,0 +1,60 @@
+
+  public HoodieBucketLayout() {
+    super();
+    partitionClass = "org.apache.hudi.table.action.commit.SparkBucketIndexPartitioner";
```
fix

minihippo commented:
@vinothchandar I addressed all comments and the failure ut is not related with this pr. Can we land this?

minihippo commented:
@hudi-bot run azure

vinothchandar commented:
Looking into the failures.

vinothchandar commented:
@minihippo I see you have added a new config class now. Can this be called `HoodieStorageLayoutConfig`  and properties all  named `hoodie.storage.layout.*`

vinothchandar commented:
@minihippo I was thinking we can name all parameters `hoodie.storage.layout..` instead, but the space curve PRs are all named `hoodie.layout.optimize` anyway. So I think its ok

vinothchandar commented:
```
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 125.714 s - in org.apache.hudi.integ.command.ITTestHoodieSyncCommand

[ERROR] Tests run: 2, Failures: 1, Errors: 0, Skipped: 1, Time elapsed: 29.812 s <<< FAILURE! - in org.apache.hudi.integ.ITTestHoodieDemo
[ERROR] org.apache.hudi.integ.ITTestHoodieDemo.testParquetDemo  Time elapsed: 29.622 s  <<< FAILURE!
org.opentest4j.AssertionFailedError: Command ([hdfs, dfsadmin, -safemode, wait]) expected to succeed. Exit (255) ==> expected: <0> but was: <255>
	at org.apache.hudi.integ.ITTestHoodieDemo.setupDemo(ITTestHoodieDemo.java:167)
	at org.apache.hudi.integ.ITTestHoodieDemo.testParquetDemo(ITTestHoodieDemo.java:107)

```

This keeps failing. Could you rebase again with latest master? want to try running the tests again

minihippo commented:
> @minihippo I was thinking we can name all parameters `hoodie.storage.layout..` instead, but the space curve PRs are all named `hoodie.layout.optimize` anyway. So I think its ok

I didn't modify the `hoodie.layout.optimize` directly, considering the history config compatibility.

vinothchandar commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieLayoutConfig.java:
```diff
@@ -0,0 +1,89 @@
+ * Storage layout related config.
+ */
+@Immutable
+@ConfigClassProperty(name = "Layout Configs",
```
We probably have to make this layout a TableConfig. i.e add it to HoodieTableConfig and persist in `hoodie.properties`, handling cases where there is going to be  tables with and without the property.

vinothchandar reviewed (approved):
@minihippo Couple more follow ups. Landing
